moved/src/move_execution/simulate.rs

```diff
@@ -0,0 +1,99 @@
+use {
+    crate::{
+        block::HeaderForExecution,
+        genesis::config::GenesisConfig,
+        move_execution::{
+            create_move_vm, create_vm_session, execute_transaction, quick_get_nonce,
+            BaseTokenAccounts,
+        },
+        primitives::{ToMoveAddress, B256},
+        types::{
+            session_id::SessionId,
+            transactions::{
+                NormalizedEthTransaction, NormalizedExtendedTxEnvelope, ScriptOrModule,
+                TransactionData, TransactionExecutionOutcome,
+            },
+        },
+        Error::InvalidTransaction,
+        InvalidTransactionCause,
+    },
+    alloy::rpc::types::TransactionRequest,
+    move_binary_format::errors::PartialVMError,
+    move_core_types::resolver::MoveResolver,
+    move_table_extension::TableResolver,
+    move_vm_runtime::module_traversal::{TraversalContext, TraversalStorage},
+    move_vm_types::gas::UnmeteredGasMeter,
+    std::time::{SystemTime, UNIX_EPOCH},
+};
+
+pub fn simulate_transaction(
+    request: TransactionRequest,
+    state: &(impl MoveResolver<PartialVMError> + TableResolver),
+    genesis_config: &GenesisConfig,
+    base_token: &impl BaseTokenAccounts,
+) -> crate::Result<TransactionExecutionOutcome> {
+    let mut tx = NormalizedEthTransaction::from(request.clone());
+    if request.from.is_some() && request.nonce.is_none() {
+        tx.nonce = quick_get_nonce(&tx.signer.to_move_address(), state);
+    }
+    let tx = NormalizedExtendedTxEnvelope::Canonical(tx);
+
+    let block_header = HeaderForExecution {
+        number: 0,
+        timestamp: SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Should get current time")
+            .as_secs(),
+        prev_randao: B256::random(),
+    };
+
+    execute_transaction(
+        &tx,
+        &B256::random(),
+        state,
+        genesis_config,
+        0,
+        base_token,
+        block_header,
+    )
+}
+
+pub fn call_transaction(
```

Comment:
This function currently skips the nonce check which is present in `execute_transaction`. We could add it in, but that would be duplicating some code. And in the future when there are more pre-/post-transaction operations (e.g. charging and refunding gas) then those operations would need to be duplicated here as well.

I understand the point of having this function separate is to enable getting return values for view calls, but I think some refactoring will be necessary to have this function reuse logic from `execute_transaction`. Probably the easiest way to do that is to factor out the pre-transaction check as another function which we can use here as well as in `execute_transaction`. Similarly, we need to make sure this function gets the gas charging functionality when it gets implemented in the future.

Reply:
Working on this.
I'll separate the transaction metadata checks from `canonical` and use it in both places.